src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,529 @@
+        }
+
+        // This method is registered by the CancellationTokenSource cts in the connect method
+        // and called by Dispose or Abort so that any open websocket connection can be closed.
+        private void AbortRequest()
+        {
+            if (State == WebSocketState.Open)
+            {
+                var closeResult = CloseAsyncCore(WebSocketCloseStatus.NormalClosure, "Connection was aborted", CancellationToken.None);
+            }
+        }
+
+        /// <summary>
+        /// Send data on <see cref="System.Net.WebSockets.ClientWebSocket"/> as an asynchronous operation.
+        /// </summary>
+        /// <returns>The async.</returns>
+        /// <param name="buffer">Buffer.</param>
+        /// <param name="messageType">Message type.</param>
+        /// <param name="endOfMessage">If set to <c>true</c> end of message.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        public override async Task SendAsync(ArraySegment<byte> buffer, WebSocketMessageType messageType, bool endOfMessage, CancellationToken cancellationToken)
+        {
+            ThrowIfNotConnected();
+
+            if (messageType != WebSocketMessageType.Binary &&
+                messageType != WebSocketMessageType.Text)
+            {
+                throw new ArgumentException($"Invalid message type: '{messageType}' specified in method 'SendAsync'.  Valid types are 'Binary' and 'Text'",
+                    nameof(messageType));
+            }
+
+            if (!endOfMessage && buffer.Array == null)
+                return;
+
+            if (!endOfMessage)
+            {
+                if (buffer.Array == null)
+                    return;
+                _writeBuffer = _writeBuffer ?? new MemoryStream();
+                _writeBuffer.Write(buffer.Array, buffer.Offset, buffer.Count);
+                return;
+            }
+            else if (_writeBuffer != null && buffer.Array != null)
+            {
+                _writeBuffer.Write(buffer.Array, buffer.Offset, buffer.Count);
+                if (!_writeBuffer.TryGetBuffer(out buffer))
+                    throw new WebSocketException(WebSocketError.NativeError);
+            }
+
+            var tcsSend = new TaskCompletionSource<bool>();
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully send or not.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            var writtenBuffer = _writeBuffer;
+            _writeBuffer = null;
+
+            using (cancellationToken.Register(() => tcsSend.TrySetCanceled()))
+            {
+                try
+                {
+                    if (messageType == WebSocketMessageType.Binary)
+                    {
+                        using (var uint8Buffer = Uint8Array.From(buffer))
+                        {
+                            _innerWebSocket!.Invoke("send", uint8Buffer);
```

Comment:
Is this synchronously sending the message?  Is it guaranteed to be fast?

Reply:
Yes this is synchronously sending the message but there is not an async method.  The send should return immediately as it supposedly buffers data but that will depend on the implementation of the browser.

Guaranteed to be as fast as the browser implementation can accept, buffer and return.  But we can not really guarantee that the implantation is fast.